_posts/2020-02-09-git-guide.markdown

```diff
@@ -0,0 +1,138 @@
+---
+layout: post
+title: git使用指南
+author: Sindweller-ch <jinghan.chen@cs2c.com.cn>
+tags: [git]
+---
+# git指南
+本文基于个人git使用经验总结而成，可供初学者上手参考，欢迎补充与指正！  
+下载安装config以及添加ssh等请自行搜索完成。
+## git仓库
+| 本地           |   fork        | 上游    |
+| :-----------: |:-------------:| :---------------:|
+| local         | origin        | upstream         |
+| 基于本地仓库修改代码 | 自己fork的仓库，拥有修改权限，可以push本地代码 | 源仓库，一般没有修改权限，所以无法直接push；拥有最新版本的代码 |
+## 本地仓库初始化相关命令
+### 需求分析：
+我要开始为开源社区做贡献了！
+### git初始化命令
+新建本地目录，在该目录下进行如下操作：
+1. 初始化本地git仓库
+`git init`
+2. 添加与之关联的fork仓库（名字可以随意，建议命名为origin）、上游仓库（名字可以随意，建议命名为upstream）  
+`git remote add origin origin_ssh_url`  
+`git remote add upstream upstream_ssh_url`  
+3. 确认关联仓库  
+`git remote -v`  
+4. 下载代码  
+`git clone origin_ssh_url`  
+ok，到这里就把github里的项目挪到本地了，下一步是开发前的准备——更新/新建分支。  
+## 更新本地代码
+### 需求分析：
+1. upstream仓库有更新，而自己fork的origin仓库和本地代码不会自动与之同步。
+2. 其他原因导致本地代码落后于upstream（或origin，极特殊情况）的代码
+- 实际上，origin/master是不需要单独更新的，因为只要将local/master更新至upstream/master，就可以在本地master的基础上新建本地分支，基于最新代码进行修改。修改完成后再push至origin上。
+### git更新本地代码命令
+1. 拉取上游代码    
+`git fetch upstream`    
+![git fetch upstream](https://img-blog.csdnimg.cn/20200209173434335.png)
+2. 同步本地分支 local/master -> upstream/master    
+`git rebase upstream/master`  
+![git rebase upstream/master](https://img-blog.csdnimg.cn/20200209173517722.png)  
+3. 此时使用`git status`查看状态    
+![git status](https://img-blog.csdnimg.cn/20200209173549834.png)  
+会发现本地代码已经同步了上游的代码，领先了origin的代码，此时已经可以开始开发了。如果你非要让origin也同步：    
+`git push origin master`    
+接下来就可以进行开发了～
+## 分支
+### 需求分析：
+1. 需要开发多个独立的功能，并以pr的方式逐一提交。
+2. 需要保留master上的代码。
+### 注：
+- 分支同样分为本地(local)分支、origin分支、upstream分支。
+- 一般在本地，与上游代码同步后，基于local/master新建本地分支，该分支中的代码继承master，且在该分支上的改动不会影响local/master中的代码。
+- 建议开发时，尽量将不同分支上开发无互相依赖的不同功能，如在branch A上开发A功能，切换到branch B上开发B功能……
+- 分支之间互不影响，但仍然要注意是否有依赖，否则会在合并时出现冲突。尽量使各功能的开发独立。
+### git分支常用命令
+`git checkout branch_name`  
+切换至branch_name分支  
+`git checkout -b new_branch`  
+新建并切换至new_branch分支  
+`git branch -a`  
+查看所有分支（标记的为所在分支）
+`git branch -D branch_name`  
+删除branch_name分支（请不要直接在该分支上做这种过河拆桥的事情）  
```

Comment:
We should make it more clear that this command only remove the local branch, neither on `origin` nor on `upstream`. If you want to remove it remotely, please use `git push origin :branch_name`.